cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,309 @@
+			return nil, fmt.Errorf("failed to retrieve firewall by ID: %s", err)
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return fw.Name == fm.workerFirewallName
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %s", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	fm.fwCache.mu.RLock()
+	targetFirewall := fm.fwCache.firewall
+	isEqual := false
+	if targetFirewall != nil && cmp.Equal(targetFirewall.InboundRules, svcInboundRules) {
+		isEqual = true
+	}
+	fm.fwCache.mu.RUnlock()
+
+	if isEqual {
+		return nil
+	}
+
+	fwSvc := fm.client.Firewalls
+
+	// Check if the target firewall ID exists but in the case that CCM first starts up and the
+	// firewall ID does not exist yet, we have to retrieve one from the API.
+	var fwID string
+	if targetFirewall == nil {
+		currentFirewall, err := fm.Get(ctx)
+		if err == nil && currentFirewall == nil {
+			// Log error then create firewall and update cache, since firewall does not exist.
+			klog.Errorf("failed to get firewall state: %s", err)
+			currentFirewall, err = fm.createFirewall(ctx, fwSvc, svcInboundRules)
+			if err != nil {
+				return err
+			}
+		}
+		// We need this ID for future Update or Create requests to the firewall service API.
+		fm.fwCache.mu.RLock()
+		fwID = currentFirewall.ID
+		fm.fwCache.mu.RUnlock()
+	} else {
+		// We need this ID for future Update or Create requests to the firewall service API.
+		fm.fwCache.mu.RLock()
+		fwID = targetFirewall.ID
+		fm.fwCache.mu.RUnlock()
+	}
+
+	// Update api firewall inbound rules with the expected inbound rules then update cache.
+	fr := &godo.FirewallRequest{
+		Name:         fm.workerFirewallName,
+		InboundRules: svcInboundRules,
+		Tags:         fm.workerFirewallTags,
+	}
+
+	// Attempt to update the firewall service with the new inbound rules.
+	currentFirewall, resp, err := fwSvc.Update(ctx, fwID, fr)
+	if err != nil {
+		if resp == nil || resp.StatusCode != http.StatusNotFound {
+			return fmt.Errorf("failed to update firewall: %v", err)
+		}
+		// Firewall does not exist anymore, so we need to create a new firewall with the
+		// updated inbound rules.
+		currentFirewall, err = fm.createFirewall(ctx, fwSvc, svcInboundRules)
+		if err != nil {
+			return fmt.Errorf("failed to add new firewall inbound rules: %s", err)
```

Comment:
How about a message like `failed to create firewall after found missing on update: %s` to make it more clear that it is about the create, and to distinguish from the earlier create we may do.